util/rowDecoder/decoder_test.go

```diff
@@ -0,0 +1,157 @@
+	decodeColsMap := make(map[int64]Column, len(cols))
+	decodeColsMap2 := make(map[int64]Column, len(cols))
+	for i := range tbl.Cols() {
+		col := tbl.Cols()[i]
+		tpExpr := Column{
+			Col: col,
+		}
+		decodeColsMap2[col.ID] = tpExpr
+		if col.GeneratedExprString != "" {
+			expr, err := expression.ParseSimpleExprCastWithTableInfo(ctx, col.GeneratedExprString, tblInfo, &col.FieldType)
+			c.Assert(err, IsNil)
+			tpExpr.GenExpr = expr
+		}
+		decodeColsMap[col.ID] = tpExpr
+	}
+	de := NewRowDecoder(tbl, decodeColsMap)
+	deWithNoGenCols := NewRowDecoder(tbl, decodeColsMap2)
+
+	timeZoneIn8, err := time.LoadLocation("Asia/Shanghai")
+	c.Assert(err, IsNil)
+	time1 := types.Time{
+		Time: types.FromDate(2019, 01, 01, 8, 01, 01, 0),
+		Type: mysql.TypeTimestamp,
+	}
+	t1 := types.NewTimeDatum(time1)
+	d1 := types.NewDurationDatum(types.Duration{
+		Duration: time.Hour + time.Second,
+	})
+
+	time2, err := time1.Add(sc, d1.GetMysqlDuration())
+	c.Assert(err, IsNil)
+	err = time2.ConvertTimeZone(timeZoneIn8, time.UTC)
+	c.Assert(err, IsNil)
+	t2 := types.NewTimeDatum(time2)
+
+	time3, err := time1.Add(sc, types.Duration{Duration: time.Hour*2 + time.Second*2})
+	c.Assert(err, IsNil)
+	err = time3.ConvertTimeZone(timeZoneIn8, time.UTC)
+	c.Assert(err, IsNil)
+	t3 := types.NewTimeDatum(time3)
+
+	testRows := []struct {
+		cols   []int64
+		input  []types.Datum
+		output []types.Datum
+	}{
+		{
+			[]int64{cols[0].ID, cols[1].ID, cols[2].ID, cols[3].ID, cols[4].ID},
+			[]types.Datum{types.NewIntDatum(100), types.NewBytesDatum([]byte("abc")), types.NewDecimalDatum(types.NewDecFromInt(1)), t1, d1},
+			[]types.Datum{types.NewIntDatum(100), types.NewBytesDatum([]byte("abc")), types.NewDecimalDatum(types.NewDecFromInt(1)), t1, d1, t2},
+		},
+		{
+			[]int64{cols[0].ID, cols[1].ID, cols[2].ID, cols[3].ID},
+			[]types.Datum{types.NewIntDatum(100), types.NewBytesDatum([]byte("abc")), types.NewDecimalDatum(types.NewDecFromInt(1)), t1},
+			[]types.Datum{types.NewIntDatum(100), types.NewBytesDatum([]byte("abc")), types.NewDecimalDatum(types.NewDecFromInt(1)), t1, types.NewDatum(nil), t3},
+		},
+		{
+			[]int64{cols[0].ID, cols[1].ID, cols[2].ID, cols[3].ID, cols[4].ID},
+			[]types.Datum{types.NewDatum(nil), types.NewDatum(nil), types.NewDatum(nil), types.NewDatum(nil), types.NewDatum(nil)},
+			[]types.Datum{types.NewDatum(nil), types.NewDatum(nil), types.NewDatum(nil), types.NewDatum(nil), types.NewDatum(nil), types.NewDatum(nil)},
+		},
+	}
+	for i, row := range testRows {
+		// test case for pk is unsigned.
+		if i > 0 {
+			c7.Flag |= mysql.UnsignedFlag
+		}
+		bs, err := tablecodec.EncodeRow(sc, row.input, row.cols, nil, nil)
+		c.Assert(err, IsNil)
+		c.Assert(bs, NotNil)
+
+		r, err := de.DecodeAndEvalRowWithMap(ctx, int64(i), bs, time.UTC, timeZoneIn8, nil)
+		c.Assert(err, IsNil)
+		// last column is pk, pk value won't store, and the pk column value is null after decode. we won't check the pk column.
```

Comment:
I got confused by `pk value won't store`, what does this mean? Thank you!

Reply:
refined comment, PTAL.